File: src/Services/IDE/Libr4.IDE.AutonomousAppGeneration.Algorithms.FSharp/Playbook/RepairPlaybookSignature.fs
module Libr4.IDE.AutonomousAppGeneration.Algorithms.Playbook.RepairPlaybookSignature

open System
open System.Collections.Generic
open System.IO
open System.Security.Cryptography
open System.Text
open System.Text.RegularExpressions

let private stopWords =
    HashSet<string>(
        [| "the"; "and"; "for"; "with"; "from"; "that"; "this"; "error"; "failed"; "cannot"; "could"; "not" |],
        StringComparer.OrdinalIgnoreCase)

let private hashRaw (raw: string) =
    let bytes = SHA256.HashData(Encoding.UTF8.GetBytes(raw.Trim().ToLowerInvariant()))
    Convert.ToHexString(bytes).ToLowerInvariant()

let private normalizeFilePattern (filePath: string option) =
    match filePath with
    | None | Some "" -> "*"
    | Some path ->
        let normalized = path.Replace('\\', '/').Trim().ToLowerInvariant()
        let fileName = Path.GetFileName(normalized)
        if String.IsNullOrWhiteSpace fileName then "*"
        else
            let ext = Path.GetExtension(fileName)
            if String.IsNullOrWhiteSpace ext then fileName else $"*{ext}"

let private extractKeywords (message: string) =
    Regex.Matches(message.ToLowerInvariant(), @"[a-z0-9_./-]{4,}")
    |> Seq.map (fun m -> m.Value)
    |> Seq.filter (fun t -> not (stopWords.Contains t))
    |> Seq.distinct
    |> Seq.sort
    |> Seq.truncate 6
    |> String.concat ","

let buildStackPattern (applicationName: string option) (languages: string[]) (frameworks: string[]) =
    match applicationName with
    | None | Some "" -> "unknown"
    | Some name ->
        let langs = String.Join(',', languages)
        let fw = String.Join(',', frameworks)
        $"{name}|{langs}|{fw}".ToLowerInvariant()

let fromError (errorType: string) (filePath: string option) (message: string) =
    let filePattern = normalizeFilePattern filePath
    let keywords = extractKeywords message
    hashRaw $"{errorType}|{filePattern}|{keywords}"

let fromBuildLog (buildLog: string) =
    let line =
        buildLog.Replace('\r', '\n').Split('\n', StringSplitOptions.RemoveEmptyEntries)
        |> Array.tryFind (fun l ->
            l.Contains("error", StringComparison.OrdinalIgnoreCase)
            || l.Contains("failed", StringComparison.OrdinalIgnoreCase)
            || l.Contains("exception", StringComparison.OrdinalIgnoreCase))
        |> Option.defaultValue buildLog
    let keywords = extractKeywords line
    hashRaw $"build|*|{keywords}"

let fromErrors
    (errors: (string * string option * string)[])
    (buildLog: string option)
    (applicationName: string option)
    (languages: string[])
    (frameworks: string[])
    =
    let stackPattern = buildStackPattern applicationName languages frameworks

    let signature =
        if errors.Length > 0 then
            let errorType, filePath, message = errors.[0]
            fromError errorType filePath message
        else
            match buildLog with
            | Some log when not (String.IsNullOrWhiteSpace log) -> fromBuildLog log
            | _ -> hashRaw "unknown|*|build_failure"

    signature, stackPattern

File: src/Services/IDE/Libr4.IDE.AutonomousAppGeneration/AgentRuntime/Playbook/RepairPlaybookSignature.cs
using Libr4.IDE.Application.AutonomousAppGeneration.Infrastructure.Algorithms;
using Libr4.IDE.Domain.AutonomousAppGeneration;

namespace Libr4.IDE.Application.AutonomousAppGeneration.AgentRuntime.Playbook;

public static class RepairPlaybookSignature
{
    public static RepairPlaybookSignatureResult FromErrors(
        IReadOnlyList<ErrorReport> errors,
        string? buildLog,
        GenerationPlan? plan)
    {
        var errorTuples = errors
            .Select(e => (e.ErrorType, e.FilePath, e.Message))
            .ToArray();

        var languages = plan?.TechStack.Languages.ToArray() ?? Array.Empty<string>();
        var frameworks = plan?.TechStack.Frameworks.ToArray() ?? Array.Empty<string>();

        var (signature, stackPattern) = FSharpAlgorithmsBridge.BuildPlaybookSignature(
            errorTuples,
            buildLog,
            plan?.ApplicationName,
            languages,
            frameworks);

        return new RepairPlaybookSignatureResult(signature, stackPattern);
    }

    public static string FromError(ErrorReport error) =>
        FSharpAlgorithmsBridge.BuildPlaybookSignature(
            new[] { (error.ErrorType, error.FilePath, error.Message) },
            null,
            null,
            Array.Empty<string>(),
            Array.Empty<string>()).Signature;

    public static string FromBuildLog(string buildLog) =>
        FSharpAlgorithmsBridge.BuildPlaybookSignature(
            Array.Empty<(string, string?, string)>(),
            buildLog,
            null,
            Array.Empty<string>(),
            Array.Empty<string>()).Signature;

    public static string BuildStackPattern(GenerationPlan? plan)
    {
        var languages = plan?.TechStack.Languages.ToArray() ?? Array.Empty<string>();
        var frameworks = plan?.TechStack.Frameworks.ToArray() ?? Array.Empty<string>();
        return FSharpAlgorithmsBridge.BuildPlaybookSignature(
            Array.Empty<(string, string?, string)>(),
            null,
            plan?.ApplicationName,
            languages,
            frameworks).StackPattern;
    }
}
